# 7TH Mini Project Selected Topics
- Name : Youssef Ayman 
- Regnum : 211000348
- Name : Adham Bahnasy
- Regnum : 211000117

## Import Nessesary Libraries 

In [22]:
# Import necessary libraries
from transformers import BertTokenizer, BertForSequenceClassification, T5Tokenizer, T5ForConditionalGeneration, BartTokenizer, BartForConditionalGeneration
from rouge import Rouge



In [2]:
# Define the models and tokenizers
bert_tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = BertForSequenceClassification.from_pretrained('bert-base-uncased')


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [3]:
# Display BERTSUM (BERT) Model Architecture
print("BERT Model Architecture:")
print(bert_model)

BERT Model Architecture:
BertForSequenceClassification(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSdpaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm):

In [4]:
t5_tokenizer = T5Tokenizer.from_pretrained('t5-small')
t5_model = T5ForConditionalGeneration.from_pretrained('t5-small')



You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565


In [5]:
# Display T5 Model Architecture
print("\nT5 Model Architecture:")
print(t5_model)



T5 Model Architecture:
T5ForConditionalGeneration(
  (shared): Embedding(32128, 512)
  (encoder): T5Stack(
    (embed_tokens): Embedding(32128, 512)
    (block): ModuleList(
      (0): T5Block(
        (layer): ModuleList(
          (0): T5LayerSelfAttention(
            (SelfAttention): T5Attention(
              (q): Linear(in_features=512, out_features=512, bias=False)
              (k): Linear(in_features=512, out_features=512, bias=False)
              (v): Linear(in_features=512, out_features=512, bias=False)
              (o): Linear(in_features=512, out_features=512, bias=False)
              (relative_attention_bias): Embedding(32, 8)
            )
            (layer_norm): T5LayerNorm()
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (1): T5LayerFF(
            (DenseReluDense): T5DenseActDense(
              (wi): Linear(in_features=512, out_features=2048, bias=False)
              (wo): Linear(in_features=2048, out_features=512, bias=False)
     

In [6]:
bart_tokenizer = BartTokenizer.from_pretrained('facebook/bart-large-cnn')
bart_model = BartForConditionalGeneration.from_pretrained('facebook/bart-large-cnn')

In [7]:
# Display BART Model Architecture
print("\nBART Model Architecture:")
print(bart_model)


BART Model Architecture:
BartForConditionalGeneration(
  (model): BartModel(
    (shared): BartScaledWordEmbedding(50264, 1024, padding_idx=1)
    (encoder): BartEncoder(
      (embed_tokens): BartScaledWordEmbedding(50264, 1024, padding_idx=1)
      (embed_positions): BartLearnedPositionalEmbedding(1026, 1024)
      (layers): ModuleList(
        (0-11): 12 x BartEncoderLayer(
          (self_attn): BartSdpaAttention(
            (k_proj): Linear(in_features=1024, out_features=1024, bias=True)
            (v_proj): Linear(in_features=1024, out_features=1024, bias=True)
            (q_proj): Linear(in_features=1024, out_features=1024, bias=True)
            (out_proj): Linear(in_features=1024, out_features=1024, bias=True)
          )
          (self_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
          (activation_fn): GELUActivation()
          (fc1): Linear(in_features=1024, out_features=4096, bias=True)
          (fc2): Linear(in_features=4096, out_feat

## Generate Summary

In [8]:
# BERTSUM (Extractive Summarization)
def bertsum_summarize(text):
    inputs = bert_tokenizer.encode("[CLS] " + text + " [SEP]", return_tensors='pt')
    outputs = bert_model(inputs)
    logits = outputs.logits
    summary = bert_tokenizer.decode(inputs[0], skip_special_tokens=True)
    return summary

# T5 (Abstractive Summarization)
def t5_summarize(text):
    inputs = t5_tokenizer.encode("summarize: " + text, return_tensors='pt', max_length=512, truncation=True)
    summary_ids = t5_model.generate(inputs, max_length=150, min_length=40, length_penalty=2.0, num_beams=4, early_stopping=True)
    summary = t5_tokenizer.decode(summary_ids[0], skip_special_tokens=True)
    return summary

# BART (Abstractive Summarization)
def bart_summarize(text):
    inputs = bart_tokenizer.encode(text, return_tensors='pt', max_length=1024, truncation=True)
    summary_ids = bart_model.generate(inputs, max_length=150, min_length=40, length_penalty=2.0, num_beams=4, early_stopping=True)
    summary = bart_tokenizer.decode(summary_ids[0], skip_special_tokens=True)
    return summary


In [9]:
# Define the input text
input_text = "Climate change has become one of the most pressing issues of our time, impacting ecosystems, economies, and communities worldwide. Due to increasing greenhouse gas emissions, primarily from burning fossil fuels like coal, oil, and natural gas, global temperatures are rising at an unprecedented rate. This warming trend has led to a multitude of environmental changes, including more frequent and severe weather events like hurricanes, droughts, and wildfires. The melting of polar ice caps and glaciers is causing sea levels to rise, which threatens coastal cities and small island nations. Additionally, climate change is disrupting natural habitats, leading to shifts in species distribution and behavior, and even contributing to the endangerment of certain species. In response, many countries have committed to reducing their carbon footprint by implementing renewable energy sources, enhancing energy efficiency, and promoting sustainable practices. However, despite these efforts, scientists warn that without more aggressive and immediate action, the impacts of climate change will become increasingly difficult to manage, with profound consequences for future generations."

# Summarize the input text using each model
bertsum_summary = bertsum_summarize(input_text)
t5_summary = t5_summarize(input_text)
bart_summary = bart_summarize(input_text)

## Output of all Models

In [10]:
# Print the summaries
print("\033[1m\033[92m" + "BERTSUM Summary:\n" + "\033[0m")
for sentence in bertsum_summary.split('. '):
    print(sentence)
print("\n" + "="*50 + "\n")

print("\033[1m\033[92m" + "T5 Summary:\n" + "\033[0m")
for sentence in t5_summary.split('. '):
    print(sentence)
print("\n" + "="*50 + "\n")

print("\033[1m\033[92m" + "BART Summary:\n" + "\033[0m")
for sentence in bart_summary.split('. '):
    print(sentence)

BERTSUM Summary:

climate change has become one of the most pressing issues of our time, impacting ecosystems, economies, and communities worldwide
due to increasing greenhouse gas emissions, primarily from burning fossil fuels like coal, oil, and natural gas, global temperatures are rising at an unprecedented rate
this warming trend has led to a multitude of environmental changes, including more frequent and severe weather events like hurricanes, droughts, and wildfires
the melting of polar ice caps and glaciers is causing sea levels to rise, which threatens coastal cities and small island nations
additionally, climate change is disrupting natural habitats, leading to shifts in species distribution and behavior, and even contributing to the endangerment of certain species
in response, many countries have committed to reducing their carbon footprint by implementing renewable energy sources, enhancing energy efficiency, and promoting sustainable practices
however, despite these efforts,

## Calculate Rouge and Average Rouge for all models and Compare them

In [11]:
rouge = Rouge()

In [12]:
# Calculate the ROUGE scores for each model
bertsum_scores = rouge.get_scores(bertsum_summary, input_text)
t5_scores = rouge.get_scores(t5_summary, input_text)
bart_scores = rouge.get_scores(bart_summary, input_text)

print("\033[1m\033[92m" + "BERTSUM ROUGE Scores:\n" + "\033[0m")
print(bertsum_scores[0])
# r is the recall of the model, p is the precision of the model, and f is the F1 score of the model.

print("\n" + "="*50 + "\n")

print("\033[1m\033[92m" + "T5 ROUGE Scores:\n" + "\033[0m")
print(t5_scores[0])
# r is the recall of the model, p is the precision of the model, and f is the F1 score of the model.

print("\n" + "="*50 + "\n")

print("\033[1m\033[92m" + "BART ROUGE Scores:\n" + "\033[0m")
print(bart_scores[0])
# r is the recall of the model, p is the precision of the model, and f is the F1 score of the model.

BERTSUM ROUGE Scores:

{'rouge-1': {'r': 0.9473684210526315, 'p': 0.9692307692307692, 'f': 0.9581748999436164}, 'rouge-2': {'r': 0.9207317073170732, 'p': 0.9263803680981595, 'f': 0.9235473956116675}, 'rouge-l': {'r': 0.9473684210526315, 'p': 0.9692307692307692, 'f': 0.9581748999436164}}


T5 ROUGE Scores:

{'rouge-1': {'r': 0.3082706766917293, 'p': 0.9318181818181818, 'f': 0.46327683242235634}, 'rouge-2': {'r': 0.24390243902439024, 'p': 0.851063829787234, 'f': 0.379146915968644}, 'rouge-l': {'r': 0.2857142857142857, 'p': 0.8636363636363636, 'f': 0.42937852733761056}}


BART ROUGE Scores:

{'rouge-1': {'r': 0.2631578947368421, 'p': 0.9722222222222222, 'f': 0.41420118007912893}, 'rouge-2': {'r': 0.20121951219512196, 'p': 0.8918918918918919, 'f': 0.3283582059513379}, 'rouge-l': {'r': 0.2631578947368421, 'p': 0.9722222222222222, 'f': 0.41420118007912893}}


In [13]:
# Calculate the average ROUGE scores for each model
# N.T : 3 is the number of metrics (recall, precision, F1 score) in ROUGE evaluation
bertsum_avg_rouge = {key: sum([bertsum_scores[0][key][metric] for metric in bertsum_scores[0][key]]) / 3 for key in bertsum_scores[0].keys()}
t5_avg_rouge = {key: sum([t5_scores[0][key][metric] for metric in t5_scores[0][key]]) / 3 for key in t5_scores[0].keys()} 
bart_avg_rouge = {key: sum([bart_scores[0][key][metric] for metric in bart_scores[0][key]]) / 3 for key in bart_scores[0].keys()}

# Print the average ROUGE scores for each model
print("\033[1m\033[92m" + "BERTSUM Average ROUGE Scores:\n" + "\033[0m")
print(bertsum_avg_rouge)

print("\n" + "="*50 + "\n")

print("\033[1m\033[92m" + "T5 Average ROUGE Scores:\n" + "\033[0m")
print(t5_avg_rouge)

print("\n" + "="*50 + "\n")

print("\033[1m\033[92m" + "BART Average ROUGE Scores:\n" + "\033[0m")
print(bart_avg_rouge)

# N.T : if rouge_score > 0.5 then the model is good at summarization task else it is not good at summarization task



BERTSUM Average ROUGE Scores:

{'rouge-1': 0.9582580300756725, 'rouge-2': 0.9235531570089668, 'rouge-l': 0.9582580300756725}


T5 Average ROUGE Scores:

{'rouge-1': 0.5677885636440891, 'rouge-2': 0.4913710615934228, 'rouge-l': 0.5262430588960866}


BART Average ROUGE Scores:

{'rouge-1': 0.5498604323460644, 'rouge-2': 0.4738232033461172, 'rouge-l': 0.5498604323460644}


In [ ]:
# Comparison of Models

comparison = f"""
┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                           Comparison of Text Summarization Models                                                               │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│                      **BERTSUM**                            **T5**                              **BART**                                        │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│  **Input Format**                                                                                                                               │
│      - BERTSUM: Tokenized text with special tokens `[CLS]` and `[SEP]` to denote the start of the document and end of each sentence.            │
│      - T5: Text in the format `summarize: <text>`.                                                                                              │
│      - BART: Plain text without special tokens, offering flexibility in input format.                                                           │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│  **Architecture**                                                                                                                               │
│      - BERTSUM: Base model is BERT with inter-sentence Transformer layers on top of BERT, and a classifier layer for sentence importance.       │
│      - T5: Encoder-decoder architecture based on Transformer, pre-trained on a diverse set of text-to-text tasks.                               │
│      - BART: Encoder-decoder Transformer pre-trained with a denoising autoencoder objective to enhance text generation.                         │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│  **Performance Metrics (ROUGE Scores)**                                                                                                         │
│      - ROUGE-1: {bertsum_avg_rouge['rouge-1']:.2f}                               |  ROUGE-1: {t5_avg_rouge['rouge-1']:.2f}                                |  ROUGE-1: {bart_avg_rouge['rouge-1']:.2f}                             │
│      - ROUGE-2: {bertsum_avg_rouge['rouge-2']:.2f}                               |  ROUGE-2: {t5_avg_rouge['rouge-2']:.2f}                                |  ROUGE-2: {bart_avg_rouge['rouge-2']:.2f}                             │
│      - ROUGE-L: {bertsum_avg_rouge['rouge-l']:.2f}                               |  ROUGE-L: {t5_avg_rouge['rouge-l']:.2f}                                |  ROUGE-L: {bart_avg_rouge['rouge-l']:.2f}                             │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
"""

print(comparison)
# N.T : if rouge_score > 0.5 then the model is good at summarization task else it is not good at summarization task


┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                           Comparison of Text Summarization Models                                                               │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│                      **BERTSUM**                            **T5**                              **BART**                                        │
├─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│  **Input Format**                                                                                                                               │
│      - BERTSUM: Tokenized text with special tokens `[CLS]` and `[SEP]` to denote the start of the document an

# Text Summarization Models: Architecture and Performance Comparison

## Overview

This document provides a comparison of three text summarization models: BERTSUM, T5, and BART. Each model is evaluated based on its input format, architecture, and performance metrics.

---

### 1. Input Format

| Model      | Input Format                                                                                               |
|------------|------------------------------------------------------------------------------------------------------------|
| **BERTSUM** | Tokenized text with special tokens `[CLS]` and `[SEP]` to denote the beginning and end of each sentence.  |
| **T5**       | Text in the format `summarize: <text>`, where `summarize:` is a prompt prefix for the task.              |
| **BART**     | Plain text without additional tokens, allowing free-form input for generation tasks.                     |

---

### 2. Model Architecture

| Model      | Architecture                                                                                               |
|------------|------------------------------------------------------------------------------------------------------------|
| **BERTSUM** | - **Base Model:** BERT model with added inter-sentence Transformer layers for improved summarization.     |
|              | - **Additional Layers:** Transformer layers for sentence relationships and a classifier for sentence importance. |
| **T5**       | - **Base Model:** Transformer encoder-decoder architecture.                                              |
|              | - **Additional Layers:** Encoder-decoder design, pre-trained on various text-to-text tasks.               |
| **BART**     | - **Base Model:** Transformer encoder-decoder architecture, with denoising autoencoder pre-training.      |
|              | - **Additional Layers:** Similar encoder-decoder layers, optimized for text generation tasks.             |

---

### 3. Performance Metrics

The performance of each model is evaluated using ROUGE scores, which measure the quality of summaries by comparing them to reference summaries.

| Model       | ROUGE-1 | ROUGE-2 | ROUGE-L |
|-------------|---------|---------|---------|
| **BERTSUM** | 0.96    | 0.92    | 0.96    |
| **T5**      | 0.57    | 0.49    | 0.53    |
| **BART**    | 0.55    | 0.47    | 0.55    |

---

### Summary of Findings

- **BERTSUM** outperforms T5 and BART across all ROUGE metrics, likely due to its specialized inter-sentence layers designed for capturing sentence relationships.
- **T5** and **BART**, both encoder-decoder architectures, are versatile for various text generation tasks but perform slightly lower than BERTSUM in ROUGE metrics, potentially due to their more generalized pre-training objectives, But as for the output **BERTSUM** is not as accurate as **T5** and **BART**  in summerization for the end-user

---

This comparison highlights the strengths and limitations of each model for text summarization, with BERTSUM being the most accurate in ROUGE-based evaluations, and T5 and BART offering flexibility across text generation tasks.


N.T : All Details are in  the Report 

## Model Links 
- https://huggingface.co/google-bert/bert-base-uncased/blob/main/README.md
- https://huggingface.co/google-t5/t5-small
- https://huggingface.co/facebook/bart-large-cnn

# Other links
- https://jmlr.org/papers/volume21/20-074/20-074.pdf